### PR TITLE
Add prefix/postfix to emitted API function definitions

### PR DESCRIPTION
**Pull request by notmgsk**
```
﻿CALLBACK-HEAVEN> (emit-api-function-definitions (compute-c-space-translation (api-group 'testy)) *standard-output* :prefix "printf(\"hi\");" :postfix "printf(\"bye\");")

int test(int a, int b) {
    int ret;
    printf("hi");
    ret = ((int (*)(int, int))(function_index_group_testy[0]))(a, b);
    printf("bye");
    return ret;}

NIL
```

### CODE REVIEW

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]: Why not `~%` at the previous `format` instead of `terpri`?

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]:
```suggestion
                (format stream "~4T~A" prefix)
```

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]: Ditto.

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]: `(format stream "~4T~A~%" postfix)`

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]: I like the hooks here for flexibility. If things were to become unwieldy we could think about using classes and letting the user of callback-heaven create a subclass implementing the desired wrappers.

**Review by jmbr — APPROVED** [review]

File: callback-heaven.lisp
Comment (notmgsk) [reply callback-heaven.lisp]: Just following the convention that was used in the rest of the file. I will break with tradition, however.

File: callback-heaven.lisp
Comment (appleby) [reply callback-heaven.lisp]: I'm also OK with keeping it simple for now, but another option would be to allow the caller to specify some set of hooks, something like `pre-body-hook`, `post-body-hook`, etc.  Then call those at appropriate points passing the `stream` and maybe some additional info like the current function name, return type, argument names and return types, etc.

**Comment by notmgsk** [comment]
That was my first thought on how to make this more generic. Happy to see it wasn’t a dumbe idea.

File: callback-heaven.lisp
Comment (jmbr) [inline callback-heaven.lisp]: `(format stream "~4Treturn ret;~%}~%~%")`

**Comment by notmgsk** [comment]
It just occurred to me that this probably isn't the best way to do this. The "expected" way to emit the library files is through `emit-library-files` which in turn eventually calls into `emit-api-function-definitions`. Of course the former doesn't take any kwargs specific to the functions it calls, and so the latter doesn't receive any kwargs.

One possible way of supporting the current implementation without having to mess with `emit-library-files` would be to use global variables in the signature:
```
(defun emit-api-function-definitions (... &key (prefix *function-definition-prefix*) (postfix *function-definition-postfix*)) 
    ...)

(let ((*function-definition-prefix "..."))
  (emit-library-files ...))
```
but there may be a better way.

**Comment by notmgsk** [comment]
Allowing `*api-function-definitions-(pre|post)fix*` to be functions:
```
CALLBACK-HEAVEN> (let ((*api-function-definition-prefix*
                         (lambda (s c f) (declare (ignore c f)) (format s "    printf(\"hi\")~%")))
                       (*api-function-definition-postfix* "printf(\"bye\");"))
                   (emit-api-function-definitions (compute-c-space-translation (api-group 'testy)) *standard-output*))

int test(int a, int b) {
    int ret;
    printf("hi")
    ret = ((int (*)(int, int))(function_index_group_testy[0]))(a, b);
    printf("bye");
    return ret;
}
```
@jmbr @appleby what do you think?

**Comment by jmbr** [comment]
This makes sense to me. It's definitely better than writing the pre- and post-code over and over again in the C code.

I'd say let's go even further and have  `emit-api-function-definitions` call `*api-function-definition-...*` if it satisfies the predicate `functionp` or print it as a string to the relevant stream otherwise.

**Comment by appleby** [comment]
Minor nitpick: can we replace `definitions` with `body` in the special var name? `*api-function-definitions-(pre|post)fix*` makes it sound to me like it gets called before/after the full function definition like

```c
// emit prefix here
int foo(void) { return 42; }
// emit postfix here
```

but maybe that's just me...

**Comment by notmgsk** [comment]
> This makes sense to me. It's definitely better than writing the pre- and post-code over and over again in the C code.
> 
> I'd say let's go even further and have `emit-api-function-definitions` call `*api-function-definition-...*` if it satisfies the predicate `functionp` or print it as a string to the relevant stream otherwise.

If I understand what you're saying -- that's what the latest commit does.

**Comment by notmgsk** [comment]
> Minor nitpick: can we replace `definitions` with `body` in the special var name? `*api-function-definitions-(pre|post)fix*` makes it sound to me like it gets called before/after the full function definition like
> 
> ```c
> // emit prefix here
> int foo(void) { return 42; }
> // emit postfix here
> ```
> 
> but maybe that's just me...

No, I agree.

File: callback-heaven.lisp
Comment (appleby) [inline callback-heaven.lisp:277]: maybe a little awkward if the `ctrans` and `f` types / accessors aren't exported

**Review by appleby — APPROVED** [review]

**Comment by appleby** [comment]
@notmgsk is this still `[don't merge]`?

**Comment by notmgsk** [comment]
No. Please give it another look.

**Comment by appleby** [comment]
> No. Please give it another look.

Looks swell to me. I already approved last week, but don't have write access to this repo, so my approval doesn't count 😞.  Should probably summon @stylewarning to get the merge merge.

**Comment by notmgsk** [comment]
@stylewarning hand out some merge permissions

**Pull request merged**
